security/src/main/scala/pl/touk/nussknacker/ui/security/api/AuthenticationResources.scala

```diff
@@ -46,18 +59,81 @@ trait AuthenticationResources extends Directives with FailFastCirceSupport {
   protected lazy val additionalRoute: Route = Directives.reject
 }
 
+object AuthenticationResources {
+
+  def apply(config: Config, classLoader: ClassLoader, sttpBackend: SttpBackend[Future, Any])(
+      implicit ec: ExecutionContext
+  ): AuthenticationResources = {
+    implicit val sttpBackendImplicit: SttpBackend[Future, Any] = sttpBackend
+    AuthenticationProvider(config, classLoader)
+      .createAuthenticationResources(config, classLoader)
+  }
+
+}
+
 trait AnonymousAccess extends Directives {
-  val anonymousUserRole: Option[String]
+  this: AuthenticationResources =>
+
+  protected def anonymousUserRole: Option[String]
+
+  protected def authenticateReally(): AuthenticationDirective[AuthenticatedUser]
+
+  protected def authenticateReally(credentials: PassedAuthCredentials): Future[Option[AuthenticatedUser]]
+
+  override protected final val authCredentialsCrypter: Crypter = AesCrypter
+
+  override def authenticationMethod(): EndpointInput[AuthCredentials] = {
+    anonymousUserRole match {
+      case Some(_) =>
+        // In WithAnonymousAccessFallback the encrypted password of the anonymous user is added as an Authorization
+        // header credentials. Here, using the same crypter, is decrypted and AuthCredentials object is created.
+        // The crypter is used to prevent the anonymous credentials from being used for explicit authentication
+        // (by passing the credentials with the credentials with the HTTP request).
```

Comment:
Can't really understand this (but I'm not really familiar with the subject so it may be just me). What would happen if we didn't use the crypter?

Reply:
the previous implementation worked like this:
```bash
curl -v "http://.../api/users/" 
```

returned `anonymous` user data.

But when you did it like this:
```bash
curl -v -u anonymous:anonymous "http://.../api/users/" 
```

you got 401. 

My solution (or hack) works like this:
when in akka http response is info that the request failed because of no credentials (and the anonymous role is configured), add the anonymous user credentials to the request and try again. 

Because of the encryption I can distinguish if our code added the credentials or the credentials come from the external world. The encryption here acts like a signature ;)